ts/client/services/events/depositFinalised.ts

```diff
@@ -0,0 +1,63 @@
+import { Event } from "ethers";
+import { CoreAPI } from "../../coreAPI";
+import { DepositPool } from "../../features/deposit";
+import { StateStorageEngine } from "../../storageEngine";
+import { ContractEventSyncer } from "./contractEventSyncer";
+
+/**
+ * Syncs DepositsFinalised events from the rollup contract
+ */
+export class DepositFinalisedEventSyncer extends ContractEventSyncer {
+    private readonly depositPool: DepositPool;
+    private readonly stateStroage: StateStorageEngine;
+
+    constructor(api: CoreAPI) {
+        super(
+            api.rollup,
+            api.rollup.filters.DepositsFinalised(null, null, null)
+        );
+        this.eventListener = this.depositFinalisedListener;
+
+        this.depositPool = api.depositPool;
+        this.stateStroage = api.l2Storage.state;
+    }
+
+    public async initialSync(
+        startBlock: number,
+        endBlock: number
+    ): Promise<void> {
+        const events = await this.getEvents(startBlock, endBlock);
+        console.info(
+            `Block ${startBlock} -- ${endBlock}\t${events.length} new deposits finalised`
+        );
+        for (const event of events) {
+            await this.handleDepositFinalised(event);
+        }
+    }
+
+    private async getEventArgs(event: Event) {
+        if (!event.args) {
+            throw new Error("DepositsFinalised event missing args");
+        }
+        const { subtreeID, depositSubTreeRoot, pathToSubTree } = event.args;
+        // TODO Properly convert types.
+        return { subtreeID, depositSubTreeRoot, pathToSubTree };
+    }
+
+    private async handleDepositFinalised(event: Event) {
+        // const { depositSubTreeRoot } = this.getEventArgs(event);
+
+        // TODO This looks like it is already being handled in newBatch event syncer via DepositHandlingStrategy:applyBatch
```

Comment:
@ChihChengLiang Can you confirm this is correct? Since the deposit batch is already being handled in the strategy with the correct subtree pop (`deposit.ts:109`) and that event (`NewBatch`) is emitted after the `DepositFinalised` event  in `submitDeposit`.

Reply:
Right the comment is correct. So we don't even need DepositFinalisedEventSyncer.

The event `DepositFinalised` always comes with the `NewBatch` event of the deposit. So we don't need an extra listener for it.

Just for extra information. In `applyBatch` we apply the state transition for the batch. There are two ways we acquire a batch, 

1. A batch might be acquired during the sync. We parse the `DepositFinalised` event from it in `handleNewBatch`.
2. A batch might come from the one the packer prepared to submit. We decided to apply the local one directly instead of sync it again on chain.